Makefile

```diff
@@ -38,25 +30,60 @@ gen-config-asserts:
 
 gen-repo-mod-replace:
 	@echo "Generating go.mod replace for repo modules"
-	cd internal/cmd/makerelative && go run ./
+	cd internal/repotools/cmd/makerelative && go run ./
 
-###################
-# Unit/CI Testing #
-###################
-build:
-	go test -tags ${ALL_TAGS} -run NONE ${SDK_ALL_PKGS}
-
-unit: verify build test-protocols test-services test-ec2imds test-credentials test-config
-	@echo "go test SDK and vendor packages"
-	@go test -tags ${UNIT_TEST_TAGS} ${SDK_ALL_PKGS}
+gen-mod-replace-local:
+	./mod_replace_local_submodules.sh `pwd` `pwd` `pwd`/../smithy-go
 
-unit-with-race-cover: verify build test-protocols test-services test-config test-credentials
-	@echo "go test SDK and vendor packages"
-	@go test -tags ${UNIT_TEST_TAGS} -race -cpu=1,2,4 ${SDK_ALL_PKGS}
 
-ci-test: generate unit-with-race-cover ci-test-generate-validate
+################
+# Unit Testing #
+################
+unit: verify build unit-test
+unit-race: verify build unit-test-race
+
+unit-test: test-modules-aws test-modules-service
+unit-test-race: test-modules-race-aws test-modules-race-service
+
+build: build-modules-.
+
+build-modules-%:
+	@# build command that uses the pattern to define the root path that the
+	@# module testing will start from. Strips off the "build-modules-" and
+	@# replaces all "_" with "/".
+	@#
+	@# e.g. build-modules-internal_protocoltest
+	cd ./internal/repotools/cmd/eachmodule \
+		&& go run . -p $(subst _,/,$(subst build-modules-,,$@)) \
+		'go test ${BUILD_TAGS} -run NONE ./...'
+
+test-modules-race-%:
+	@# Test command that uses the pattern to define the root path that the
+	@# module testing will start from. Strips off the "test-modules-race-" and
+	@# replaces all "_" with "/".
+	@#
+	@# e.g. test-modules-race-internal_protocoltest
+	cd ./internal/repotools/cmd/eachmodule \
+		&& go run . -p $(subst _,/,$(subst test-modules-race-,,$@)) \
+		'go test ${BUILD_TAGS} -run NONE ./...' \
+		'go test -timeout=1m ${UNIT_TEST_TAGS} -race -cpu=1,2,4 ./...'
+
+test-modules-%:
+	@# Test command that uses the pattern to define the root path that the
+	@# module testing will start from. Strips off the "test-modules-" and
+	@# replaces all "_" with "/".
+	@#
+	@# e.g. test-modules-internal_protocoltest
+	cd ./internal/repotools/cmd/eachmodule \
+		&& go run . -p $(subst _,/,$(subst test-modules-,,$@)) \
+		'go test ${BUILD_TAGS} -run NONE ./...' \
```

Comment:
Just a styling perspective, but since the `-run` argument takes a regex should we just pass `'^$'` as the match to actually truely specify none?

Reply:
Sure can switch to that, `NONE` was just a convention I'd been using the other can be used instead